### PR TITLE
feat：add messageTask modular

### PR DESCRIPTION
**Pull request by wzh425**

### CODE REVIEW

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/ReceiverItemDto.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Masa.Mc.Contracts.Admin.Dtos.MessageTasks;
+
+public class ReceiverItemDto
+{
+    public string DataId { get; set; } = string.Empty;
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/ReceiverItemDto.cs:5]: SubjectId

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/Validator/MessageTaskCreateUpdateDtoValidator.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+public class MessageTaskCreateUpdateDtoValidator : AbstractValidator<MessageTaskCreateUpdateDto>
+{
+    public MessageTaskCreateUpdateDtoValidator()
+    {
+        RuleFor(input => input.ChannelId).Required();
+        RuleFor(input => input.EntityId).Required().When(x => x.EntityType == MessageEntityType.Template);
+        RuleFor(input => input.EntityType).IsInEnum();
+        RuleFor(input => input.ReceiverType).IsInEnum();
+        RuleFor(input => input.Receivers).Must(x => x.Items != null && x.Items.Count > 0).When(x => x.ReceiverType == ReceiverType.Assign);
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/Validator/MessageTaskCreateUpdateDtoValidator.cs:11]: x.Items?.Any()

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/Validator/SendMessageTaskInputValidator.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Masa.Mc.Contracts.Admin.Dtos.MessageTasks.Validator;
+
+public class SendMessageTaskInputValidator : AbstractValidator<SendMessageTaskInput>
+{
+    public SendMessageTaskInputValidator()
+    {
+        RuleFor(input => input.ReceiverType).IsInEnum();
+        RuleFor(input => input.Receivers).Must(x => x.Items != null && x.Items.Count > 0).When(x => x.ReceiverType == ReceiverType.Assign);
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/Validator/SendMessageTaskInputValidator.cs:8]: Any

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/ReceiverGroups/ReceiverImportDto.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿namespace Masa.Mc.Contracts.Admin.Dtos.ReceiverGroups;
+
+[Importer(HeaderRowIndex = 2, IsDisableAllFilter = true)]
+public class ReceiverImportDto
+{
+    /// <summary>
+    ///  昵称
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/ReceiverGroups/ReceiverImportDto.cs:7]: Modify comment

File: src/Infrastructure/Masa.Mc.Infrastructure.ObjectExtending/ObjectExtending/ExtensionPropertyHelper.cs [hunk 5]
```diff
@@ -1,4 +1,6 @@
-﻿namespace Masa.Mc.Infrastructure.ObjectExtending.ObjectExtending;
+﻿using System.Text.Json;
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.ObjectExtending/ObjectExtending/ExtensionPropertyHelper.cs:1]: global using

File: src/Services/Masa.Mc.Service/Infrastructure/EntityFrameworkCore/McDbContext.cs [hunk 6]
```diff
@@ -18,6 +20,16 @@ public McDbContext(MasaDbContextOptions<McDbContext> options) : base(options)
     protected override void OnModelCreatingExecuting(ModelBuilder builder)
     {
         base.OnModelCreatingExecuting(builder);
+
+        builder.Entity<AppChannel>(b =>
+        {
+            b.ToTable(MCConsts.DbTablePrefix + "Channels", MCConsts.DbSchema);
+            b.Property(x => x.Code).IsRequired().HasMaxLength(64).HasColumnName(nameof(Channel.Code));
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Infrastructure/EntityFrameworkCore/McDbContext.cs:27]: default value is property name

Comment (wzh425) [reply src/Services/Masa.Mc.Service/Infrastructure/EntityFrameworkCore/McDbContext.cs:27]: Shared entity needs HasColumnName

**Review by MayueCif — APPROVED** [review]

**Comment by MayueCif** [comment]
@doddgu LGTM

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/MessageTaskHistoryDto.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Dtos.MessageTasks;
+
+public class MessageTaskHistoryDto : AuditEntityDto<Guid, Guid>
+{
+    public Guid MessageTaskId { get; set; }
+
+    public ReceiverType ReceiverType { get; set; }
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/MessageTaskHistoryDto.cs:10]: ReceiverTypes

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/ReceiverItemDto.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+    public string DisplayName { get; set; } = string.Empty;
+
+    public string Avatar { get; set; } = string.Empty;
+
+    public string PhoneNumber { get; set; } = string.Empty;
+
+    public string Email { get; set; } = string.Empty;
+
+    public MessageTaskReceiverType Type { get; set; }
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/ReceiverItemDto.cs:18]: MessageTaskReceiverTypes

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/SendingRuleDto.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Dtos.MessageTasks;
+
+public class SendingRuleDto
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/SendingRuleDto.cs:6]: ing?

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/DisableMessageTaskInput.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Dtos.MessageTasks;
+
+public class DisableMessageTaskInput
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/DisableMessageTaskInput.cs:6]: Input -> InputDto

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTemplates/SmsTemplateSynchroInput.cs [hunk 11]
```diff
@@ -1,4 +1,7 @@
-﻿namespace Masa.Mc.Contracts.Admin.Dtos.MessageTemplates;
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Dtos.MessageTemplates;
 
 public class SmsTemplateSynchroInput
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTemplates/SmsTemplateSynchroInput.cs:6]: Synchro -> Sync

File: src/Contracts/Masa.Mc.Contracts.Admin/Enums/Channels/ChannelType.cs [hunk 12]
```diff
@@ -1,4 +1,7 @@
-﻿namespace Masa.Mc.Contracts.Admin.Enums.Channels;
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Enums.Channels;
 
 public enum ChannelType
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Enums/Channels/ChannelType.cs:6]: ChannelTypes

File: src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageEntityType.cs [hunk 13]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Enums.MessageTasks;
+
+public enum MessageEntityType
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageEntityType.cs:6]: MessageEntityTypes

File: src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageTaskHistoryStatus.cs [hunk 14]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Enums.MessageTasks;
+
+public enum MessageTaskHistoryStatus
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageTaskHistoryStatus.cs:6]: MessageTaskHistoryStatuses

File: src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageTaskReceiverType.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Enums.MessageTasks;
+
+public enum MessageTaskReceiverType
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageTaskReceiverType.cs:6]: MessageTaskReceiverTypes

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTemplates/MessageTemplateCreateUpdateDto.cs [hunk 16]
```diff
@@ -18,13 +21,14 @@ public MessageTemplateCreateUpdateDto()
...
     public string Sign { get; set; } = string.Empty;
-    public MessageTemplateStatus Status { get; set; } = MessageTemplateStatus.Normal;
-    public MessageTemplateAuditStatus AuditStatus { get; set; } = MessageTemplateAuditStatus.Adopt;
+    public MessageTemplateStatues Status { get; set; } = MessageTemplateStatues.Normal;
+    public MessageTemplateAuditStatues AuditStatus { get; set; } = MessageTemplateAuditStatues.Adopt;
     public DateTime? AuditTime { get; set; }
     public DateTime? InvalidTime { get; set; }
     public string AuditReason { get; set; } = string.Empty;
     public int TemplateType { get; set; }
     public long DayLimit { get; set; }
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTemplates/MessageTemplateCreateUpdateDto.cs:30]: per day

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs [hunk 17]
```diff
@@ -1,7 +1,90 @@
...
+    /// </summary>
+    /// <param name="source">The collection</param>
+    /// <param name="items">Item to check and add</param>
+    /// <typeparam name="T">Type of the items in the collection</typeparam>
+    /// <returns>Returns the added items.</returns>
+    public static IEnumerable<T> AddIfNotContains<T>(this ICollection<T> source, IEnumerable<T> items)
+    {
+        var addedItems = new List<T>();
+
+        foreach (var item in items)
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs:48]: like this source.AddRange(item.Where(a=>source.Contains(a)));

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs [hunk 18]
```diff
@@ -1,7 +1,90 @@
...
+    /// <summary>
+    /// Removes all items from the collection those satisfy the given <paramref name="predicate"/>.
+    /// </summary>
+    /// <typeparam name="T">Type of the items in the collection</typeparam>
+    /// <param name="source">The collection</param>
+    /// <param name="predicate">The condition to remove the items</param>
+    /// <returns>List of removed items</returns>
     public static IList<T> RemoveAll<T>(this ICollection<T> source, Func<T, bool> predicate)
     {
         var items = source.Where(predicate).ToList();
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs:90]: source.ToList().RemoveAll(predicate);

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/DictionaryExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+    /// <summary>
+    /// Gets a value from the dictionary with given key. Returns default value if can not find.
+    /// </summary>
+    /// <param name="dictionary">Dictionary to check and get</param>
+    /// <param name="key">Key to find the value</param>
+    /// <typeparam name="TKey">Type of the key</typeparam>
+    /// <typeparam name="TValue">Type of the value</typeparam>
+    /// <returns>Value if found, default if can not found.</returns>
+    public static TValue GetOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key)
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/DictionaryExtensions.cs:54]: repeat 40 line

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/EnumerableExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+public static class EnumerableExtensions
+{
+    /// <summary>
+    /// Concatenates the members of a constructed <see cref="IEnumerable{T}"/> collection of type System.String, using the specified separator between each member.
+    /// This is a shortcut for string.Join(...)
+    /// </summary>
+    /// <param name="source">A collection that contains the strings to concatenate.</param>
+    /// <param name="separator">The string to use as a separator. separator is included in the returned string only if values has more than one element.</param>
+    /// <returns>A string that consists of the members of values delimited by the separator string. If values has no members, the method returns System.String.Empty.</returns>
+    public static string JoinAsString(this IEnumerable<string> source, string separator)
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/EnumerableExtensions.cs:18]: this extension method whether or not necessary

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/StringExtensions.cs [hunk 21]
```diff
@@ -0,0 +1,545 @@
...
+            ? Regex.Replace(str, "[a-z][A-Z]", m => m.Value[0] + " " + char.ToLower(m.Value[1]))
+            : Regex.Replace(str, "[a-z][A-Z]", m => m.Value[0] + " " + char.ToLowerInvariant(m.Value[1]));
+    }
+
+    /// <summary>
+    /// Converts given PascalCase/camelCase string to kebab-case.
+    /// </summary>
+    /// <param name="str">String to convert.</param>
+    /// <param name="useCurrentCulture">set true to use current culture. Otherwise, invariant culture will be used.</param>
+    public static string ToKebabCase(this string str, bool useCurrentCulture = false)
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/StringExtensions.cs:291]: public static string ToKebabCase(this string str)
        {
            if (string.IsNullOrEmpty(str))
            {
                return str;
            }
            return Regex.Replace(str, @"(\B[A-Z])", "-$1").ToLower();
        }

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Helper/TypeHelper.cs [hunk 22]
```diff
@@ -0,0 +1,340 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Infrastructure.Common.Helper;
+
+public static class TypeHelper
+{
+    private static readonly HashSet<Type> FloatingTypes = new HashSet<Type>
+        {
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Common/Helper/TypeHelper.cs:9]: format

File: src/Infrastructure/Masa.Mc.Infrastructure.Email/EmailSenderBase.cs [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+            mail.From = new MailAddress(
+                options.DefaultFromAddress,
+                options.DefaultFromDisplayName,
+                Encoding.UTF8
+                );
+        }
+
+        if (mail.HeadersEncoding == null)
+        {
+            mail.HeadersEncoding = Encoding.UTF8;
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Email/EmailSenderBase.cs:60]: mail.HeadersEncoding = mail.HeadersEncoding??Encoding.UTF8;

File: src/Infrastructure/Masa.Mc.Infrastructure.Email/Infrastructure/OptionsResolve/Contributors/ConfigurationOptionsResolveContributor.cs [hunk 24]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Infrastructure.Email.Infrastructure.OptionsResolve.Contributors;
+
+public class ConfigurationOptionsResolveContributor : IEmailOptionsResolveContributor
+{
+    public const string ContributorName = "Configuration";
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Email/Infrastructure/OptionsResolve/Contributors/ConfigurationOptionsResolveContributor.cs:8]: const name rule all upper

File: src/Infrastructure/Masa.Mc.Infrastructure.Email/Infrastructure/OptionsResolve/IEmailOptionsResolver.cs [hunk 25]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Infrastructure.Email.Infrastructure.OptionsResolve;
+
+public interface IEmailOptionsResolver
+{
+    Task<IEmailOptions> ResolveAsync();
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.Email/Infrastructure/OptionsResolve/IEmailOptionsResolver.cs:8]: Resolve？

File: src/Services/Masa.Mc.Service/Domain/MessageTemplates/EventHandler/MessageTemplateAuditStatusChangedEventHandler.cs [hunk 26]
```diff
@@ -13,15 +16,15 @@ public MessageTemplateAuditStatusChangedEventHandler(IMessageTemplateRepository
     public async Task MessageTemplateStatusChangedToApproved(MessageTemplateAuditStatusChangedToApprovedDomainEvent @event)
     {
         var entity = await _repository.FindAsync(x => x.Id == @event.TemplateId);
-        entity.SetAuditStatus(MessageTemplateAuditStatus.Adopt, @event.Remarks);
+        entity.SetAuditStatus(MessageTemplateAuditStatues.Adopt, @event.Remarks);
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Domain/MessageTemplates/EventHandler/MessageTemplateAuditStatusChangedEventHandler.cs:19]: Statues?Statuses

File: src/Web/Masa.Mc.Web.Admin/Pages/MessageTasks/Modules/MessageSendingRules.razor.cs [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+
+    [Parameter]
+    public bool ReadOnly { get; set; }
+
+    [Parameter]
+    public EventCallback<DateTime?> SendTimeChanged { get; set; }
+
+    private bool _datePickersShow;
+    private bool _timePickersShow;
+    public DateOnly? sendingDate;
```
Comment (MayueCif) [inline src/Web/Masa.Mc.Web.Admin/Pages/MessageTasks/Modules/MessageSendingRules.razor.cs:22]: public filed?

File: src/Web/Masa.Mc.Web.Admin/Shared/MessageType.cs [hunk 28]
```diff
@@ -1,4 +1,7 @@
-﻿using System;
```
Comment (MayueCif) [inline src/Web/Masa.Mc.Web.Admin/Shared/MessageType.cs:1]: delete MessageType.cs

**Review by MayueCif — COMMENTED** [review]
modify by conversation

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs (hunk 17 above)
Comment (wzh425) [reply src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs:48]: ICollection does not have AddRange method

File: src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs (hunk 18 above)
Comment (wzh425) [reply src/Infrastructure/Masa.Mc.Infrastructure.Common/Extensions/CollectionExtensions.cs:90]: ToList Will cause EF save

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/EnabledMessageTaskInput.cs [hunk 29]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Dtos.MessageTasks;
+
+public class EnabledMessageTaskInput
```
[the remark this answers is not in the document]
Comment (wzh425) [reply src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/EnabledMessageTaskInput.cs]: post body parameter

**Pull request merged**